feat(signals): Video-first session summarization

Twixes opened the pull request:
## Problem

We've started with session summaries based on just the events, but we've seen the reliability of such description to be lower than when we use an actual session video.

Accurate descriptions of what the user genuinely experienced are essential, as they will allow us to identify real clusters of user experiences, letting us propose meaningful engineering tasks with improvements.

## Changes

This adds a new pathway of creating a single session summary, implemented as a sequence of Temporal activities (behind flag `max-session-summarization-video-as-base`):

1. `export_session_video_activity` – exports a full video of the session via ExportedAsset creation (if session already has a video ExportedAsset, this is skipped)
2. `upload_video_to_gemini_activity` – queries ExportedAsset for the relevant video, and uploads it to the Gemini Files API
3. `analyze_video_segment_activity` – each run is a different 15s chunk of the full session video analyzed by Gemini (with events present in the 15s span included for context)
4. `consolidate_video_segments_activity` – concats the LLM-generated descriptions of all the 15s chunks, and runs them through one last LLM call to segment the complete session transcript into user journey segments
5. `embed_and_store_segments_activity` – queues every user journey segment to `embedding-worker`, storing every segment in an embedded format for later clustering 
6. `store_video_session_summary_activity` – takes the summary created so far, and serializes it into our good old friend SingleSessionSummary, in the same format as classic summaries

(these files are prefixed with `a` for "activity", like `a1_export_session_video`, because reflecting the order is so helpful, but also Python requires module names to start with a letter)

### Example result (Hedgebox dummy app)

<img width="674" height="609" alt="Based on video (+ events)" src="https://github.com/user-attachments/assets/11a12b0b-bbc0-4c55-9737-68efc6a12d43" />


### Which existing building blocks are used here

- No new _workflows_ introduced here, only an alternate branch of activities for the existing `SummarizeSingleSessionWorkflow`, which also results in the same `SessionSummarySerializer` output as existing session summarization. This branching is done on `inputs.video_validation_enabled == "full"`.
- The same Redis caching used to pass heavy state between activites.
- Video exports kick off the well-known VideoExportWorkflow.
- For embeddings, we hand off to `embedding-worker`, which handles the embedding processing async (this is valuable for later clustering, but not for immediate consumption).

### What is **missing**

- _Critical_: Support for stream mode of single session summarization (`SummarizeSingleSessionStreamWorkflow`) – this gets kicked off directly in the session replay UI, and I've not yet rejigged it in this PR as this one's _large_ already; means "Summarize this session" button errors out
- _Important_: Support for longer sessions – because we currently do `playback_speed` of 1.0 in the full session export, long sessions take proportionally long to export, and somewhere the summarization workflow timeout right now is 600 s, so sessions longer 10 minutes fail 😿 
- _Worth testing_: Gemini 3 for consolidation of segments (no blockers, just haven't compared yet)
- _Worth testing_: Description of where we are in the segment transcription prompt (but if the user navigates, pageview events do get included)
- _Worth testing_: Storing embeddings of the raw 15s segments (currently only the consolidated journey segments get embedded)
- _Worth testing_: Adding metadata on what features were used to the consolidated segments
- _Worth testing_: Mapping to the existing `abondonment` or `confusion` fields of summaries is there, but feels meh, I'm not sure it's useful as is

## How did you test this code?

Mostly manual testing via the new `python manage.py team_signals__summarize_single_session <session_id>` command.

Currently no tests covering the new branches. Do bash my lack of tests.

Twixes committed:
feat(signals): Video-first session summarization

Twixes committed:
feat(signals): Video-first session summarization

Twixes committed:
Add segment consolidation step

Twixes committed:
Add segment consolidation step

Twixes committed:
Remove test code

Twixes committed:
Remove test code

Twixes committed:
Split up activities into one per file

Twixes committed:
Split up activities into one per file

Twixes committed:
Add test command

Twixes committed:
Add test command

Twixes committed:
Use shared util for duration extraction

Twixes committed:
Use shared util for duration extraction

Twixes committed:
Rename files for etra clarity

Twixes committed:
Rename files for etra clarity

Twixes committed:
Update nodes.py

Twixes committed:
Update nodes.py

Twixes committed:
Skip sessions too short and skip first 2s

Twixes committed:
Skip sessions too short and skip first 2s

Twixes commented on ee/hogai/videos/utils.py:
```diff
@@ -0,0 +1,17 @@
+    for track in media_info.tracks:
+        if track.track_type == "General":
+            # Convert ms to seconds, ceil to avoid grey "not-rendered" frames at the start
+            return math.ceil(track.duration / 1000.0)
```
Q @sortafreel: I copied this verbatim from the existing code, but I'm not sure if it's actually intended to ceil here? Seems like if we wanted to skip, we'd floor. And either way, this should not be needed, as we have a more intentional `SESSION_VIDEO_RENDERING_DELAY` offset

Twixes committed:
Refactor prompts and improve tracing

Twixes committed:
Refactor prompts and improve tracing

Twixes committed:
Align final output with original schema

Twixes committed:
Align final output with original schema

Twixes committed:
DRY the code

Twixes committed:
DRY the code

Twixes committed:
Address Greptile comments

Twixes committed:
Address Greptile comments

Twixes committed:
Update test_summarize_session_group.py

Twixes committed:
Update test_summarize_session_group.py

Twixes committed:
Update a2_upload_video_to_gemini.py

Twixes committed:
Update a2_upload_video_to_gemini.py

Twixes committed:
Update summarize_session.py

Twixes committed:
Update summarize_session.py

Twixes committed:
Update test_session_summaries.py

Twixes committed:
Update test_session_summaries.py

sortafreel committed:
Merge branch 'master' into video-based-session-summaries

sortafreel commented on ee/api/session_summaries.py:
```diff
@@ -93,8 +93,18 @@ def _validate_input(self, request: Request) -> tuple[list[str], datetime, dateti
 
-    def _if_video_validation_enabled(self, user: User) -> bool | None:
-        # Check if the summaries should be validated with videos
+    def _determine_video_validation_enabled(self, user: User) -> bool | Literal["full"] | None:
```
Don't like 3 return types even one bit, but understand the reasoning, ok to keep it.

sortafreel commented on posthog/temporal/ai/session_summary/summarize_session.py:
```diff
@@ -384,6 +393,115 @@ async def run(self, inputs: SingleSessionSummaryInputs) -> None:
 
 
+async def ensure_llm_single_session_summary(inputs: SingleSessionSummaryInputs):
+    if inputs.video_validation_enabled == "full":
```
Nitpick: No need to keep the code nested. If not full - start and return regular activity, no need for else afterward.

sortafreel commented on posthog/temporal/ai/session_summary/summarize_session.py:
```diff
@@ -384,6 +393,115 @@ async def run(self, inputs: SingleSessionSummaryInputs) -> None:
+        # Calculate segment specs based on video duration
+        # (We ignore the first couple of seconds, as they often include malformed frames)
+        analysis_duration = uploaded_video.duration - SESSION_VIDEO_RENDERING_DELAY
+        num_segments = int(analysis_duration / SESSION_VIDEO_CHUNK_DURATION_S)  # Number of segments is floored
```
Seems valid?

sortafreel commented on posthog/temporal/ai/session_summary/activities/a1_export_session_video.py:
```diff
@@ -0,0 +1,124 @@
+    """Export full session video and return ExportedAsset ID"""
+    try:
+        # Check for existing exported asset for this session
+        existing_asset = (
```
In the ideal world we would also want to track uploaded video file id, so we can:
- Avoid uploading it again, if we are restarting the flow
- Remove it when the video is processed to not hit Gemini File API limits

Not sure how to do it smoothly though. I'll add a couple of TODO comments for the future usage.

sortafreel replied:
Sure, not a blocker.

sortafreel committed:
fix: Don't block the loop with time.

sortafreel committed:
fix: Don't block the loop with time.

sortafreel committed:
fix: Increase timeout for video export, as we use 1x.

sortafreel committed:
fix: Increase timeout for video export, as we use 1x.

sortafreel committed:
chore: Add TODO.

sortafreel committed:
chore: Add TODO.

sortafreel committed:
fix: Log exceptions with tag.

sortafreel committed:
fix: Log exceptions with tag.

sortafreel commented on posthog/temporal/ai/session_summary/activities/a1_export_session_video.py:
```diff
@@ -0,0 +1,124 @@
+            session_id=inputs.session_id,
+            team=team,
+        )
+        if not metadata:
```
Updated to log exceptions with tag in this file. I think it's worth attaching the tag to all logs to unify, so it's easier to find.

sortafreel commented on posthog/temporal/ai/session_summary/activities/a2_upload_video_to_gemini.py:
```diff
@@ -0,0 +1,95 @@
+            )
+            # Wait for file to be ready
+            while uploaded_file.state and uploaded_file.state.name == "PROCESSING":
+                time.sleep(0.5)  # Gotta do polling sadly
```
Fixed.

sortafreel commented on posthog/temporal/ai/session_summary/activities/a3_analyze_video_segment.py:
```diff
@@ -0,0 +1,328 @@
+                    session_start_time_str=llm_input.session_start_time_str,
+                )
+
+                if events_in_range:
```
What would happen if no events in range?

sortafreel commented on posthog/temporal/ai/session_summary/activities/a3_analyze_video_segment.py:
```diff
@@ -0,0 +1,328 @@
+        team_name = (await Team.objects.only("name").aget(id=inputs.team_id)).name
+
+        # Analyze with Gemini using video_metadata to specify the time range
+        client = genai.Client(api_key=settings.GEMINI_API_KEY)
```
`from posthog.ai.gemini import AsyncClient`?

sortafreel commented on posthog/temporal/ai/session_summary/activities/a3_analyze_video_segment.py:
```diff
@@ -0,0 +1,328 @@
+
+        # Analyze with Gemini using video_metadata to specify the time range
+        client = genai.Client(api_key=settings.GEMINI_API_KEY)
+        response = client.models.generate_content(
```
Same comment, looks like blocking activity.

sortafreel commented on posthog/temporal/ai/session_summary/activities/a3_analyze_video_segment.py:
```diff
@@ -0,0 +1,328 @@
+- Describe what's happening in the video as a list of salient moments
+- Highlight what features were used, and what the user was doing with them
+- Note any problems, errors, confusion, or friction the user experienced
+- If tracked events show exceptions ($exception_types, $exception_values), identify them in your analysis
```
`identify` is a bit fluid. I would suggest stating something in the lines of "validate if they happened in the video", as without it LLM could trust event data. The point is to specifically allow to ignore events data if video doesn't match it. The same for the prompt below.

sortafreel committed:
chore: Add TODO.

sortafreel committed:
chore: Add TODO.

sortafreel commented on posthog/temporal/ai/session_summary/activities/a6_store_video_session_summary.py:
```diff
@@ -0,0 +1,496 @@
+    }
+
+
+def _convert_video_segments_to_session_summary_fallback(
```
Not sure about this part. If something failed this miserably - we probably need just to fail completely, raise an exception, and then update the solution to not fail, instead of trying to return something minimally usable. Up to the discussion though.

Twixes replied:
Left over from a previous approach – I removed the whole fallback, as it should indeed not happen, and if it does something went quite wrong

sortafreel commented on posthog/temporal/ai/session_summary/types/video.py:
```diff
@@ -0,0 +1,82 @@
+
+    segments: list[ConsolidatedVideoSegment]
+    session_outcome: VideoSessionOutcome
+    segment_outcomes: list[dict]  # [{segment_index: int, success: bool, summary: str}]
```
Question: If you created dataclasses for all the stuff above, why not for the segment outcome?

sortafreel replied:
We can also add validation on that class.

Twixes replied:
Just omitted, created now!

sortafreel commented on posthog/temporal/ai/session_summary/activities/a3_analyze_video_segment.py:
```diff
@@ -0,0 +1,329 @@
+    """Format seconds as MM:SS or HH:MM:SS"""
+    minutes = int(seconds // 60)
+    secs = int(seconds % 60)
+    return f"{minutes:02d}:{secs:02d}"
```
It seems to return only MM:SS, but not HH?

sortafreel commented on posthog/temporal/ai/session_summary/types/video.py:
```diff
@@ -0,0 +1,82 @@
+    """Specification for a segment of video to analyze"""
+
+    segment_index: int
+    start_time: float  # Seconds from start of video
```
Nitpick: You probably do it somewhere, but I would suggest adding check that end time is actually later than start time, on the class level.

Twixes replied:
Added a Pydantic validator ✅

sortafreel commented on posthog/temporal/ai/session_summary/summarize_session.py:
```diff
@@ -384,6 +393,115 @@ async def run(self, inputs: SingleSessionSummaryInputs) -> None:
+        ]
+
+        # Activity 3: Analyze all segments in parallel
+        segment_tasks = [
```
I would suggest adding `Semaphore` with limit of 10 or smth, as we can easily hit RPM/TPM.

sortafreel commented on posthog/management/commands/team_signals__summarize_single_session.py:
```diff
@@ -0,0 +1,102 @@
+            import traceback
+
+            self.stdout.write(traceback.format_exc())
+            raise
```
Do we need to do any cleanup?

sortafreel reviewed (changes requested):
Lots of blocking comments, approved as a matter of habbit 😆

sortafreel committed:
Merge branch 'master' into video-based-session-summaries

sortafreel committed:
Merge branch 'video-based-session-summaries' of github.com:PostHog/posthog into video-based-session-summaries

sortafreel committed:
Merge branch 'master' into video-based-session-summaries

sortafreel committed:
Merge branch 'master' into video-based-session-summaries

sortafreel committed:
Merge branch 'video-based-session-summaries' of github.com:PostHog/posthog into video-based-session-summaries

Twixes committed:
Remove unused `asset_id`

Twixes committed:
Add flush

Twixes committed:
Handle None duration

Twixes committed:
Addressing type notes

Twixes committed:
Add upload timeout

Twixes committed:
Add comment to each activity file

Twixes committed:
Reduce nesting

Twixes commented on posthog/temporal/ai/session_summary/summarize_session.py:
```diff
@@ -384,6 +393,115 @@ async def run(self, inputs: SingleSessionSummaryInputs) -> None:
+            user_distinct_id_to_log=inputs.user_distinct_id_to_log,
+            team_id=inputs.team_id,
+            redis_key_base=inputs.redis_key_base,
+            model_to_use="gemini-2.5-flash",  # Default model for video analysis
```
Updated to use `DEFAULT_VIDEO_UNDERSTANDING_MODEL`. Note: This was previously `gemini-2.5-flash-preview-09-2025`, I changed to the GA version of the model i.e. `gemini-2.5-flash`

Twixes committed:
Use DEFAULT_VIDEO_UNDERSTANDING_MODEL

Twixes committed:
Make skipping logic for too-short videos more robust

Twixes committed:
Use AsyncClient

This requires a `posthoganalytics.ai` upgrade

Twixes committed:
Consolidate empty `raw_segments` handling

Twixes commented on posthog/temporal/ai/session_summary/activities/a4_consolidate_video_segments.py:
```diff
@@ -0,0 +1,189 @@
+    - Success/failure of the overall session
+    - Per-segment outcomes (success, confusion, abandonment, failures)
+    """
+    if not raw_segments:
```
Yeah, it already raises an error, but a level up in `ensure_llm_single_session_summary`. Moved that check here to have all the logic in the relevant activities, making this clearer.

Twixes committed:
Remove needless _convert_video_segments_to_session_summary_fallback

Twixes committed:
Merge branch 'master' into video-based-session-summaries

Twixes committed:
Add VideoSegmentOutcome dataclass

Twixes committed:
Clarify _format_timestamp_as_mm_ss

Twixes committed:
Be Pydantic

Twixes committed:
Speed up video playback a bit

Twixes commented on posthog/temporal/ai/session_summary/activities/a1_export_session_video.py:
```diff
@@ -0,0 +1,124 @@
+                "timestamp": 0,  # Start from beginning
+                "filename": filename,
+                "duration": session_duration,
+                "playback_speed": 1.0,  # Normal speed
```
I played with the speed param – I think we can safely do 2.0 right now, but not more, as CSS animations do _not_ get sped up right now. From a quick look, I haven't found a way to speed up those animations in Chrome/Playwright, but I might have missed it.
It looks like by far the biggest win is in ignoring inactivity, so my take is: we go with 2.0 playback, but then in a follow-up PR we prune inactivity. The active periods are typically not that long

Twixes committed:
Validate end_time vs. start_time

Twixes committed:
Restore try:catch:

Twixes committed:
Tune logging levels

Twixes committed:
Improve segment error handling

Twixes committed:
Optimize team_name

Twixes committed:
Fix stray inline import

Twixes committed:
Revert "Optimize team_name"

This reverts commit 36d4ebc198294ffe2bb155e2dcb20323f74e146a.

Twixes committed:
Address AI review points

Twixes committed:
Make mypy happy

Twixes committed:
Improve "Static"

Twixes commented on posthog/temporal/ai/session_summary/activities/a3_analyze_video_segment.py:
```diff
@@ -0,0 +1,328 @@
+        # Parse response into segments
+        segments = []
+
+        if response_text.lower() == "static":
```
I think this branch actually shouldn't be needed, as the LLM output should always be in the timestamp-based format. Updated this slightly.

Twixes committed:
Make example a bit more varied

Twixes committed:
Merge branch 'master' into video-based-session-summaries

Twixes committed:
Merge branch 'master' into video-based-session-summaries

Twixes committed:
Allow VideoExportWorkflow duplicates

oliverb123 reviewed (approved):
:shipit:

Twixes committed:
Fix metadata in embedding records

Twixes committed:
Merge branch 'master' into video-based-session-summaries

Twixes committed:
Merge branch 'master' into video-based-session-summaries

Twixes committed:
Merge branch 'master' into video-based-session-summaries

github-actions[bot] committed:
test(storybook): update UI snapshots

sortafreel commented on ee/hogai/session_summaries/constants.py:
```diff
@@ -42,12 +42,12 @@
 EXPIRES_AFTER_DAYS = 90  # How long to store the videos used for validation
 DEFAULT_VIDEO_EXPORT_MIME_TYPE = "video/webm"
-DEFAULT_VIDEO_UNDERSTANDING_MODEL = "gemini-2.5-flash-preview-09-2025"
+DEFAULT_VIDEO_UNDERSTANDING_MODEL = "gemini-2.5-flash"
```
I would switch to 3, as we would switch to it anyway if we want to extract metadata/OCR/etc., as it 3 seems (vibe) to deal with it way better than 2.5.

sortafreel commented on posthog/temporal/ai/session_summary/activities/a1_export_session_video.py:
```diff
@@ -0,0 +1,144 @@
+        session_duration = metadata["duration"]  # duration is in seconds
+
+        # Check if session is too short for summarization - note: this is different from the video duration, but probs close enough
+        if session_duration * 1000 < MIN_SESSION_DURATION_FOR_SUMMARY_MS:
```
In the worst case we would have 1s session (- cutoff from both sides). I think `MIN_SESSION_DURATION_FOR_SUMMARY_MS` should be larger, like, at least 15s?

sortafreel commented on posthog/temporal/ai/session_summary/activities/a6_store_video_session_summary.py:
```diff
@@ -0,0 +1,496 @@
+        hours, minutes, seconds = int(parts[0]), int(parts[1]), int(parts[2])
+        return (hours * 3600 + minutes * 60 + seconds) * 1000
+    else:
+        raise ValueError(f"Invalid timestamp format: {timestamp_str}")
```
`signals_type="session-summaries"`, I find it helpful to tag all logs like that, so it's searchable in Loki.

Twixes replied:
Added a bunch

Twixes committed:
Use gemini-3-flash-preview for video

Twixes committed:
Limit analyze_video_segment_activity to 100 at once

Twixes committed:
Add bunch of signals_type="session-summaries"

Twixes committed:
Avoid fetching Team for each segment analysis

Twixes committed:
Improve event comparison

Twixes committed:
Set separate MIN_SESSION_DURATION_FOR_VIDEO_SUMMARY_S

Twixes committed:
Remove redundant dict update

Twixes committed:
Delete personProfileLogic.ts

Twixes committed:
Make exceptions blocking (or not)

Twixes committed:
Merge branch 'master' into video-based-session-summaries

Twixes committed:
Update posthog/temporal/ai/session_summary/activities/a1_export_session_video.py

Co-authored-by: graphite-app[bot] <96075541+graphite-app[bot]@users.noreply.github.com>

Twixes committed:
FIx unused import

Twixes committed:
Fix mypy's issue

Twixes committed:
Kill unreliable test_flag_with_behavioural_cohorts test

https://posthog.slack.com/archives/C08JQTX5RRP/p1768587686488789?thread_ts=1768586523.141779&cid=C08JQTX5RRP
